nebuchadnezzar/nebu/cli/pre_assemble.py

```diff
@@ -0,0 +1,224 @@
+"""Inject / modify metadata for book CNXML from git"""
+
+from pathlib import Path
+from datetime import timezone
+from typing import Optional
+import os
+import logging
+
+import click
+from lxml import etree
+from git import Repo
+
+from ._common import common_params
+from ..utils import re_first_or_default, unknown_progress
+from ..models.book_container import BookContainer
+from ..models.path_resolver import PathResolver
+from ..parse import NSMAP as CNXML_NSMAP
+from ..xml_utils import open_xml
+
+
+NS_MDML = CNXML_NSMAP["md"]
+NS_COLLXML = CNXML_NSMAP["col"]
+NS_CNXML = CNXML_NSMAP["c"]
+GIT_SHA_PREFIX_LEN = 7
+
+
+logger = logging.getLogger("nebuchadnezzar")
+
+
+def check_for_existing_metadata(xml_doc, tags, sourcefile):
+    existing = xml_doc.xpath(
+        "|".join(f"//md:{tag}" for tag in tags), namespaces={"md": NS_MDML}
+    )
+    assert (
+        len(existing) == 0
+    ), f"Unexpectedly found one of ({', '.join(tags)}) in {sourcefile}"
+
+
+def add_metadata_entries(xml_doc, new_metadata, md_namespace):
+    """Insert metadata entries from dictionairy into document"""
+    metadata = xml_doc.xpath("//x:metadata", namespaces={"x": md_namespace})[0]
+
+    for tag, value in new_metadata.items():
+        element = etree.Element(f"{{{NS_MDML}}}{tag}")
+        element.text = value
+        element.tail = "\n"
+        metadata.append(element)
+
+
+def determine_book_version(reference, repo, commit):
+    """Determine the book version string given a reference, a git repo, and
+    a target a commit"""
+    # We want to check if the provided reference is a tag that matches the
```

Comment:
I'm not sure if I understand the purpose of this but when would we care about the tag for anything other than checkout (during `step-fetch`)?

It seems that using a tag could mean that repeated builds of the same book would result in different output (e.g. if the user built a book before a tag was created and then again after the tag was created)